src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/conf/WriteYamlStep.java

```diff
@@ -0,0 +1,173 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.utility.steps.conf;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.TaskListener;
+import org.jenkinsci.plugins.pipeline.utility.steps.shaded.org.yaml.snakeyaml.DumperOptions;
+import org.jenkinsci.plugins.pipeline.utility.steps.shaded.org.yaml.snakeyaml.Yaml;
+import org.jenkinsci.plugins.workflow.steps.*;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import javax.inject.Inject;
+import java.io.*;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.Map;
+
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Writes a yaml file from the workspace.
+ *
+ * @author Javier DELGADO &lt;witokondoria@gmail.com&gt;.
+ */
+public class WriteYamlStep extends AbstractStepImpl {
+
+	private String file = null;
+	private Object data = null;
+
+	@DataBoundConstructor
+	public WriteYamlStep() {
+	}
+
+	/**
+	 * Name of the yaml file to write.
+	 *
+	 * @return file name
+	 */
+	public String getFile() {
+		return file;
+	}
+
+	/**
+	 * Name of the yaml file to write.
+	 *
+	 * @param file file name
+	 */
+	@DataBoundSetter
+	public void setFile(String file) {
+	    this.file = file;
+	}
+
+	/**
+	 * An Object containing data to be saved.
+	 *
+	 * @return data to save as yaml
+	 */
+	public Object getData() {
+	    return data;
+	}
+
+	/**
+	 * An Object representing the elements to write.
+	 *
+	 * @param data to parse
+	 */
+	@DataBoundSetter
+	public void setData(Object data) {
+	    this.data = data;
+	}
+
+	@Extension
+	public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+		public DescriptorImpl() {
+			super(Execution.class);
+		}
+
+		@Override
+		public String getFunctionName() {
+			return "writeYaml";
+		}
+
+		@Override
+		public String getDisplayName() {
+			return "Write a yaml from an object.";
+		}
+
+		@Override
+		public Step newInstance(Map<String, Object> arguments) throws Exception {
+			WriteYamlStep step = new WriteYamlStep();
```

Comment:
I doubt this is needed for this step, skip it and just put `file` and `data` as required params in the `@DataBoundConstructor`.
The special handling here that you've seen in other steps are because they have an _either or_ relation that can't be expressed in a better way.